### PR TITLE
JR2-197 Implement JobCenter REST endpoint

### PR DESCRIPTION
**Pull request by varusakov**

### CODE REVIEW

File: src/main/java/ch/admin/seco/service/reference/service/impl/JobCenterServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+public class JobCenterServiceImpl implements JobCenterService {
+
+    private final Logger log = LoggerFactory.getLogger(JobCenterServiceImpl.class);
+
+    private final JobCenterRepository jobCenterRepository;
+    private final JobCenterMapper jobCenterMapper;
+
+    private final BiFunction<Set<Address>, Language, Set<Address>> filterAddressesByLanguage = (addresses, language) -> addresses.stream()
+        .filter(address -> address.getLanguage().equals(language))
+        .collect(Collectors.toSet());
```
Comment (michael-wirth) [inline src/main/java/ch/admin/seco/service/reference/service/impl/JobCenterServiceImpl.java]: use `findFirst` instead there is no need for another collection and rename function to findAddressByLanguage

File: src/main/java/ch/admin/seco/service/reference/service/impl/JobCenterServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+     * @param language address language
+     * @return JobCenterDto with only one address
+     */
+    @Override
+    @Transactional(readOnly = true)
+    public Optional<JobCenterDto> findJobCenterByCode(String code, Language language) {
+        log.debug("Request to search JobCenter by code : {} and language : {}", code, language);
+        return jobCenterRepository.findOneByCode(code)
+            .map(jobCenter -> {
+                Set<Address> filteredAddresses = filterAddresses(jobCenter.getAddresses(), language);
```
Comment (michael-wirth) [inline src/main/java/ch/admin/seco/service/reference/service/impl/JobCenterServiceImpl.java]: move this function into the mapper. all the mapper logic (even selecting of the proper address) should be part of the mapper

File: src/main/java/ch/admin/seco/service/reference/service/impl/JobCenterServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                jobCenter.setAddresses(filteredAddresses);
+                return jobCenterMapper.jobCenterToDto(jobCenter);
```
Comment (michael-wirth) [inline src/main/java/ch/admin/seco/service/reference/service/impl/JobCenterServiceImpl.java]: use`jobCenterMapper.jobCenterToDto(jobCenter, language);` here

File: src/main/java/ch/admin/seco/service/reference/service/dto/mapper/JobCenterMapper.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    private final AddressMapper addressMapper;
+
+    public JobCenterMapper(AddressMapper addressMapper) {
+        this.addressMapper = addressMapper;
+    }
+
+    public JobCenterDto jobCenterToDto(JobCenter jobCenter) {
+        AddressDto addressDto = jobCenter.getAddresses().stream().findFirst()
+            .map(addressMapper::addressToDto)
+            .orElse(null);
```
Comment (michael-wirth) [inline src/main/java/ch/admin/seco/service/reference/service/dto/mapper/JobCenterMapper.java]: the default should be the `de` German address

File: src/main/resources/config/liquibase/changelog/data/REFERENCESERVICE_PUBLIC_JOB_CENTER.csv [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+e3b47c38-cd0d-4110-8f6e-3c2f10efb991,BEAL0,orp.bienne@vol.be.ch,+41 31 635 35 60,
+86801eb0-6aaa-4cbf-bbbc-532391ccc5c3,BEA20,orp.tavannes@vol.be.ch,+41 31 636 17 00,
+4a162076-5808-443d-9b08-60a456de81b1,VDW10,orp.payerne@vd.ch,+41 26 557 31 30,
+7c65d8a9-5d40-4a66-b260-f3c14c7722aa,BSA71,reto.frey@bs.ch,+41 61 267 50 91,+41 61 267 50 80
+ea3dbe5d-34aa-4dc7-a659-6068c5cfe36d,AGAJ0,pfortearbeitsmarkt@ag.ch,+41 62 765 95 95,+41 62 765 95 96
+c7031265-b2d8-49c2-8904-ccfd3dd7e247,NEA25,orpvdt@ne.ch,+41 32 889 61 50,+41 32 889 61 59
+748297b3-6cee-42c8-92f4-c4d3691b4c12,BSA42,josef.birchler@bs.ch,061 267 87 62,
+8d6afa20-fd5b-4339-a910-bf32b556c126,BSA82,jobroom.rav@bs.ch,+41 61 267 50 33,+41 61 267 50 52
+7836aa42-4663-4425-b584-ff8d037868a6,BSA72,reto.frey@bs.ch,+41 61 267 50 91,+41 61 267 50 80
+42dc884c-c24d-4033-be24-2f5c27787210,AGA72,rav.brugg@ag.ch,+41 56 460 90 40,+41 56 460 90 48
```
Comment (michael-wirth) [inline src/main/resources/config/liquibase/changelog/data/REFERENCESERVICE_PUBLIC_JOB_CENTER.csv]: always have a new line at EOF

File: src/main/java/ch/admin/seco/service/reference/domain/valueobject/Address.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+    @Column(name = "house_number")
+    @NotNull
+    private String houseNumber;
+
+    @Column(name = "zip_code")
+    @NotNull
+    @Min(1000)
+    @Max(9999)
+    private int zipCode;
```
Comment (michael-wirth) [inline src/main/java/ch/admin/seco/service/reference/domain/valueobject/Address.java]: Use String instead.
We used String for ZipCode also in other entities and ZipCode is by definition never truly numeric.There are some discussion on the internet regarding this topic. So let's prefer String for `zipCode`

File: src/main/java/ch/admin/seco/service/reference/service/dto/AddressDto.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+    private String name;
+
+    private String city;
+
+    private String street;
+
+    private String houseNumber;
+
+    private int zipCode;
```
Comment (michael-wirth) [inline src/main/java/ch/admin/seco/service/reference/service/dto/AddressDto.java]: Also use `String` here

File: src/main/java/ch/admin/seco/service/reference/service/dto/mapper/JobCenterMapper.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+            .code(jobCenter.getCode())
+            .email(jobCenter.getEmail())
+            .phone(jobCenter.getPhone())
+            .fax(jobCenter.getFax())
+            .address(addressDto);
+
+    }
+
+    private Optional<Address> findAddressByLanguage(Set<Address> addresses, Language language) {
+        Optional<Address> address = findAddressByLanguage.apply(addresses, language);
```
Comment (michael-wirth) [inline src/main/java/ch/admin/seco/service/reference/service/dto/mapper/JobCenterMapper.java]: You already get back an `Optional`. just use it and add `.orElse(...)` as we know there should be an Address this method should return an `Address` - not Optional

Anyway there might be a cyclic dependency when there is not the requested NOR a german Address

Comment (varusakov) [reply src/main/java/ch/admin/seco/service/reference/service/dto/mapper/JobCenterMapper.java]: `.orElse()` requires `Address` instance but `findByLanguage` returns `Optional<Address>`. We could call `Optional.get` to fetch german address and return it from `orElse` but from my POV it's not safe. I would propose to left it as is to take into account situation when german address not exists.

**Pull request merged**